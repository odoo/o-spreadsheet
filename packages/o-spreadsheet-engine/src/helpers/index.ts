export * from "./misc";
export * from "./numbers";
export * from "./references";
export * from "./uuid";
