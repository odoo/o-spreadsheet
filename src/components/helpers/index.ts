export * from "@odoo/o-spreadsheet-engine/components/helpers/css";
