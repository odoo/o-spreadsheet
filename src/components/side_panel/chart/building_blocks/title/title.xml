<templates>
  <t t-name="o-spreadsheet.ChartTitle">
    <t t-set="placeholder">Add a title</t>
    <Section class="'o-chart-title pb-0'">
      <t t-set-slot="title">
        <t t-if="props.name" t-esc="props.name"/>
        <t t-else="">Title</t>
      </t>
      <input
        type="text"
        class="o-input o-optional"
        t-att-value="props.title"
        t-on-change="updateTitle"
        t-att-placeholder="placeholder"
      />
      <Section class="'pt-0 px-0'">
        <div class="o-chart-title-designer position-relative d-flex align-items-center">
          <span
            class="o-menu-item-button o-hoverable-button"
            title="Bold"
            t-att-class="{active: props.style.bold}"
            t-on-click="(ev) => this.props.toggleBold()">
            <span>
              <t t-call="o-spreadsheet-Icon.BOLD"/>
            </span>
          </span>
          <span
            class="o-menu-item-button o-hoverable-button"
            title="Italic"
            t-att-class="{active: props.style.italic}"
            t-on-click="(ev) => this.props.toggleItalic()">
            <span>
              <t t-call="o-spreadsheet-Icon.ITALIC"/>
            </span>
          </span>
          <span
            class="o-menu-item-button o-hoverable-button"
            title="Horizontal alignment"
            t-on-click="(ev) => this.toggleDropdownTool('horizontalChartAlignTool', ev)">
            <span>
              <t t-if="props.style.align === 'center'" t-call="o-spreadsheet-Icon.ALIGN_CENTER"/>
              <t t-elif="props.style.align === 'right'" t-call="o-spreadsheet-Icon.ALIGN_RIGHT"/>
              <t t-else="" t-call="o-spreadsheet-Icon.ALIGN_LEFT"/>
            </span>
            <t t-call="o-spreadsheet-Icon.CARET_DOWN"/>
          </span>
          <div
            class="o-dropdown-content position-absolute top-100 start-0 bg-white"
            t-if="state.activeTool === 'horizontalChartAlignTool'"
            t-on-click.stop="">
            <div class="o-dropdown-line d-flex">
              <span
                class="o-menu-item-button o-hoverable-button"
                t-att-class="{active: props.style.align === 'left'}"
                title="Left"
                t-on-click="(ev) => this.updateAlignment('left')">
                <span>
                  <t t-call="o-spreadsheet-Icon.ALIGN_LEFT"/>
                </span>
              </span>
              <span
                class="o-menu-item-button o-hoverable-button"
                t-att-class="{active: props.style.align === 'center'}"
                title="Center"
                t-on-click="(ev) => this.updateAlignment('center')">
                <span>
                  <t t-call="o-spreadsheet-Icon.ALIGN_CENTER"/>
                </span>
              </span>
              <span
                class="o-menu-item-button o-hoverable-button"
                t-att-class="{active: props.style.align === 'right'}"
                title="Right"
                t-on-click="(ev) => this.updateAlignment('right')">
                <span>
                  <t t-call="o-spreadsheet-Icon.ALIGN_RIGHT"/>
                </span>
              </span>
            </div>
          </div>
          <ColorPickerWidget
            currentColor="props.style.color"
            toggleColorPicker="(ev) => this.toggleDropdownTool('fillChartColorTool', ev)"
            showColorPicker="state.activeTool === 'fillChartColorTool'"
            onColorPicked.bind="onColorPicked"
            title="fill_color"
            icon="'o-spreadsheet-Icon.TEXT_COLOR'"
            class="'o-hoverable-button o-menu-item-button'"
          />
        </div>
      </Section>
    </Section>
  </t>
</templates>
