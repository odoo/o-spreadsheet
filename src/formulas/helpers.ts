import { functionRegistry } from "../functions";
import { AST, ASTFuncall, iterateAstNodes, parseTokens } from "./parser";
import { Token } from "./tokenizer";

const functions = functionRegistry.content;

export function isExportableToExcel(tokens: Token[]): boolean {
  try {
    const nonExportableFunctions = iterateAstNodes(parseTokens(tokens)).filter(
      (ast) => ast.type === "FUNCALL" && !functions[ast.value.toUpperCase()]?.isExported
    );
    return nonExportableFunctions.length === 0;
  } catch (error) {
    return false;
  }
}

export function getFunctionsFromTokens(tokens: Token[], functionNames: string[]) {
  // Parsing is an expensive operation, so we first check if the
  // formula contains one of the function names
  if (!tokens.some((t) => t.type === "SYMBOL" && functionNames.includes(t.value.toUpperCase()))) {
    return [];
  }
  let ast: AST | undefined;
  try {
    ast = parseTokens(tokens);
  } catch {
    return [];
  }
  return getFunctionsFromAST(ast, functionNames);
}

function getFunctionsFromAST(ast: AST, functionNames: string[]) {
  return iterateAstNodes(ast)
    .filter((node) => node.type === "FUNCALL" && functionNames.includes(node.value.toUpperCase()))
    .map((node: ASTFuncall) => ({
      functionName: node.value.toUpperCase(),
      args: node.args,
    }));
}
