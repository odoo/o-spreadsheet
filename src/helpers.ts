import { Zone } from "./model/state";

/**
 * This files contains a small collection of useful helpers. They are mostly
 * concerned with the following topics:
 *
 * - coordinate conversions
 * - zone manipulation
 * - miscellaneous
 */

//------------------------------------------------------------------------------
// Coordinate
//------------------------------------------------------------------------------

/**
 * Convert a (col) number to the corresponding letter.
 *
 * Examples:
 *     0 => 'A'
 *     25 => 'Z'
 *     26 => 'AA'
 *     27 => 'AB'
 */
export function numberToLetters(n: number): string {
  if (n < 26) {
    return String.fromCharCode(65 + n);
  } else {
    return numberToLetters(Math.floor(n / 26) - 1) + numberToLetters(n % 26);
  }
}

/**
 * Convert a string (describing a column) to its number value.
 *
 * Examples:
 *     'A' => 0
 *     'Z' => 25
 *     'AA' => 26
 */
function lettersToNumber(letters: string): number {
  let result = 0;
  const l = letters.length;
  for (let i = 0; i < l; i++) {
    let n = letters.charCodeAt(i) - 65 + (i < l - 1 ? 1 : 0);
    result += n * 26 ** (l - i - 1);
  }
  return result;
}

/**
 * Convert a "XC" coordinate to cartesian coordinates.
 *
 * Examples:
 *   A1 => [0,0]
 *   B3 => [1,2]
 *
 * Note: it also accepts lowercase coordinates
 */
export function toCartesian(xc: string): [number, number] {
  xc = xc.toUpperCase();
  const [m, letters, numbers] = xc.match(/([A-Z]*)([0-9]*)/)!;
  if (m !== xc) {
    throw new Error(`Invalid cell description: ${xc}`);
  }
  const col = lettersToNumber(letters);
  const row = parseInt(numbers, 10) - 1;
  return [col, row];
}

/**
 * Convert from cartesian coordinate to the "XC" coordinate system.
 *
 * Examples:
 *   [0,0] => A1
 *   [1,2] => B3
 */
export function toXC(col: number, row: number): string {
  return numberToLetters(col) + String(row + 1);
}

export function zoneToXC(zone: Zone): string {
  const topLeft = toXC(zone.left, zone.top);
  const botRight = toXC(zone.right, zone.bottom);

  if (topLeft != botRight) {
    return topLeft + ":" + botRight;
  }

  return topLeft;
}

//------------------------------------------------------------------------------
// Zones
//------------------------------------------------------------------------------

/**
 * Compute the union of two zones. It is the smallest zone which contains the
 * two arguments.
 */
export function union(z1: Zone, z2: Zone): Zone {
  return {
    top: Math.min(z1.top, z2.top),
    left: Math.min(z1.left, z2.left),
    bottom: Math.max(z1.bottom, z2.bottom),
    right: Math.max(z1.right, z2.right)
  };
}

/**
 * Two zones are equal if they represent the same area, so we clearly cannot use
 * reference equality.
 */
export function isEqual(z1: Zone, z2: Zone): boolean {
  return (
    z1.left === z2.left && z1.right === z2.right && z1.top === z2.top && z1.bottom === z2.bottom
  );
}

/**
 * Return true if two zones overlap, false otherwise.
 */
export function overlap(z1: Zone, z2: Zone): boolean {
  if (z1.bottom < z2.top || z2.bottom < z2.top) {
    return false;
  }
  if (z1.right < z2.left || z2.right < z1.left) {
    return false;
  }
  return true;
}

export function isInside(col: number, row: number, zone: Zone): boolean {
  const { left, right, top, bottom } = zone;
  return col >= left && col <= right && row >= top && row <= bottom;
}

//------------------------------------------------------------------------------
// Miscellaneous
//------------------------------------------------------------------------------

/**
 * Stringify an object, like JSON.stringify, except that the first level of keys
 * is ordered.
 */
export function stringify(obj: any): string {
  return JSON.stringify(obj, Object.keys(obj).sort());
}
