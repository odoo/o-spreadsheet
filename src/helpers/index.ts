export * from "./color";
export * from "./coordinates";
export * from "./dates";
export * from "./edge_scrolling";
export * from "./format";
export * from "./misc";
export * from "./numbers";
export * from "./range";
export * from "./references";
export * from "./search";
export * from "./sheet";
export * from "./uuid";
export * from "./zones";
