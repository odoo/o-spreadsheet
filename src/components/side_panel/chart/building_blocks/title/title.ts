import { Component, useExternalListener, useState } from "@odoo/owl";
import { Color, SpreadsheetChildEnv, TitleDesign } from "../../../../../types";
import { ColorPickerWidget } from "../../../../color_picker/color_picker_widget";
import { css } from "../../../../helpers";
import { Section } from "../../../components/section/section";

css/* scss */ `
  .o-chart-title-designer {
    > span {
      height: 30px;
    }

    .o-menu-item-button.active {
      background-color: #e6f4ea;
      color: #188038;
    }

    .o-dropdown-content {
      overflow-y: auto;
      overflow-x: hidden;
      padding: 2px;
      z-index: 100;
      box-shadow: 1px 2px 5px 2px rgba(51, 51, 51, 0.15);

      .o-dropdown-line {
        > span {
          padding: 4px;
        }
      }
    }
  }
`;

interface Props {
  title?: string;
  updateTitle: (title: string) => void;
  name?: string;
  toggleItalic?: () => void;
  toggleBold?: () => void;
  updateAlignment?: (string) => void;
  updateColor?: (Color) => void;
  style: TitleDesign;
}

export interface ChartTitleState {
  activeTool: string;
}

export class ChartTitle extends Component<Props, SpreadsheetChildEnv> {
  static template = "o-spreadsheet.ChartTitle";
  static components = { Section, ColorPickerWidget };
  static props = {
    title: { type: String, optional: true },
    updateTitle: Function,
    name: { type: String, optional: true },
    toggleItalic: { type: Function, optional: true },
    toggleBold: { type: Function, optional: true },
    updateAlignment: { type: Function, optional: true },
    updateColor: { type: Function, optional: true },
    style: { type: Object, optional: true },
  };
  static defaultProps = {
    title: "",
  };
  openedEl: HTMLElement | null = null;

  setup() {
    useExternalListener(window, "click", this.onExternalClick);
  }

  state = useState({
    activeTool: "",
  });

  updateTitle(ev: InputEvent) {
    this.props.updateTitle((ev.target as HTMLInputElement).value);
  }

  toggleDropdownTool(tool: string, ev: MouseEvent) {
    const isOpen = this.state.activeTool === tool;
    this.closeMenus();
    this.state.activeTool = isOpen ? "" : tool;
    this.openedEl = isOpen ? null : (ev.target as HTMLElement);
  }

  /**
   * TODO: This is clearly not a goot way to handle external click, but
   * we currently have no other way to do it ... Should be done in
   * another task to handle the fact we want only one menu opened at a
   * time with something like a menuStore ?
   */
  onExternalClick(ev: MouseEvent) {
    if (this.openedEl === ev.target) {
      return;
    }
    this.closeMenus();
  }

  onColorPicked(color: Color) {
    this.props.updateColor?.(color);
    this.closeMenus();
  }

  updateAlignment(aligment: "left" | "center" | "right") {
    this.props.updateAlignment?.(aligment);
    this.closeMenus();
  }

  closeMenus() {
    this.state.activeTool = "";
    this.openedEl = null;
  }
}
