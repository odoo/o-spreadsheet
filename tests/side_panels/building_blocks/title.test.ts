import { ChartTitle } from "../../../src/components/side_panel/chart/building_blocks/title/title";
import { click, setInputValueAndTrigger } from "../../test_helpers/dom_helper";
import { mountComponentWithPortalTarget } from "../../test_helpers/helpers";

let fixture: HTMLElement;

async function mountChartTitle(props: ChartTitle["props"]) {
  ({ fixture } = await mountComponentWithPortalTarget(ChartTitle, { props }));
}

describe("Chart title", () => {
  test("Can render a chart title component", async () => {
    await mountChartTitle({
      title: "My title",
      updateTitle: () => {},
      style: {},
    });
    expect(fixture).toMatchSnapshot();
  });

  test("Can render a chart title component with default title prop if not provided", async () => {
    await mountChartTitle({
      updateTitle: () => {},
      style: {},
    });

    const input = fixture.querySelector("input")!;
    expect(input.value).toBe("");
  });

  test("Update is called when title is changed, not on input", async () => {
    const updateTitle = jest.fn();
    await mountChartTitle({
      title: "My title",
      updateTitle,
      style: {},
    });
    const input = fixture.querySelector("input")!;
    expect(input.value).toBe("My title");
    await setInputValueAndTrigger(input, "My new title", "onlyInput");
    expect(updateTitle).toHaveBeenCalledTimes(0);
    input.dispatchEvent(new Event("change"));
    expect(updateTitle).toHaveBeenCalledTimes(1);
  });

  test("UpdateColor is called when title color is changed", async () => {
    const updateColor = jest.fn();
    await mountChartTitle({
      title: "My title",
      updateTitle: () => {},
      style: {},
      updateColor,
    });
    expect(updateColor).toHaveBeenCalledTimes(0);
    await click(fixture, ".o-color-picker-button");
    await click(fixture, ".o-color-picker-line-item[data-color='#EFEFEF'");
    expect(updateColor).toHaveBeenCalledWith("#EFEFEF");
  });

  test.each(["Left", "Center", "Right"])(
    "UpdateAlignment is called when alignment is changed",
    async (alignment: string) => {
      const updateAlignment = jest.fn();
      await mountChartTitle({
        title: "My title",
        updateTitle: () => {},
        style: {},
        updateAlignment,
      });
      expect(updateAlignment).toHaveBeenCalledTimes(0);
      await click(fixture, ".o-menu-item-button[title='Horizontal alignment']");
      await click(fixture, `.o-menu-item-button[title='${alignment}']`);
      expect(updateAlignment).toHaveBeenCalledWith(alignment.toLowerCase());
    }
  );

  test("ToggleBold is called when clicking on bold button", async () => {
    const toggleBold = jest.fn();
    await mountChartTitle({
      title: "My title",
      updateTitle: () => {},
      style: {},
      toggleBold,
    });
    expect(toggleBold).toHaveBeenCalledTimes(0);
    await click(fixture, ".o-menu-item-button[title='Bold']");
    expect(toggleBold).toHaveBeenCalledTimes(1);
  });

  test("ToggleItalic is called when clicking on italic button", async () => {
    const toggleItalic = jest.fn();
    await mountChartTitle({
      title: "My title",
      updateTitle: () => {},
      style: {},
      toggleItalic,
    });
    expect(toggleItalic).toHaveBeenCalledTimes(0);
    await click(fixture, ".o-menu-item-button[title='Italic']");
    expect(toggleItalic).toHaveBeenCalledTimes(1);
  });
});
