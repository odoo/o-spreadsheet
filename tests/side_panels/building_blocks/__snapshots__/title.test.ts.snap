// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chart title Can render a chart title component 1`] = `
<div>
  <div
    class="o-spreadsheet"
  >
    <div
      class="o-section o-chart-title pb-0"
    >
      <div
        class="o-section-title"
      >
        
        Title
      </div>
      
      <input
        class="o-input o-optional"
        placeholder="Add a title"
        type="text"
      />
      <div
        class="o-section pt-0 px-0"
      >
        
        <div
          class="o-chart-title-designer position-relative d-flex align-items-center"
        >
          <span
            class="o-menu-item-button o-hoverable-button"
            title="Bold"
          >
            <span>
              <svg
                class="o-icon"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M13.5 6.5C13.5 4.57 11.93 3 10 3H4.5v12h6.25c1.79 0 3.25-1.46 3.25-3.25 0-1.3-.77-2.41-1.87-2.93.83-.58 1.37-1.44 1.37-2.32M9.5 5c.83 0 1.5.67 1.5 1.5S10.33 8 9.5 8h-2V5h2m-2 8v-3H10c.83 0 1.5.67 1.5 1.5S10.83 13 10 13H7.5"
                  fill="currentColor"
                />
              </svg>
            </span>
          </span>
          <span
            class="o-menu-item-button o-hoverable-button"
            title="Italic"
          >
            <span>
              <svg
                class="o-icon"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M7 3v2h2.58l-3.66 8H3v2h8v-2H8.42l3.66-8H15V3"
                  fill="currentColor"
                />
              </svg>
            </span>
          </span>
          <span
            class="o-menu-item-button o-hoverable-button"
            title="Horizontal alignment"
          >
            <span>
              
              
              <svg
                class="o-icon align-left"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M2 16h10v-2H2v2M12 6H2v2h10V6M2 2v2h14V2H2m0 10h14v-2H2v2"
                  fill="currentColor"
                />
              </svg>
              
            </span>
            <div
              class="o-icon fa-small"
            >
              <i
                class="fa fa-caret-down"
              />
            </div>
            
          </span>
          
          <div
            class="o-color-picker-widget"
          >
            <span
              class="o-color-picker-button o-hoverable-button o-menu-item-button"
            >
              <span
                style="border-bottom-style: hidden"
              >
                <svg
                  class="o-icon"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M10 1H8L3.5 13h2l1.12-3h4.75l1.12 3h2L10 1ZM7.38 8 9 3.67 10.62 8H7.38"
                    fill="currentColor"
                  />
                </svg>
              </span>
            </span>
            
          </div>
          
        </div>
        
      </div>
      
    </div>
  </div>
</div>
`;
